src/transformers/models/whisper/modeling_whisper.py

```diff
@@ -680,6 +680,10 @@ def _init_weights(self, module):
             module.weight.data.normal_(mean=0.0, std=std)
             if module.padding_idx is not None:
                 module.weight.data[module.padding_idx].zero_()
+        elif isinstance(module, WhisperEncoder):
```

Comment:
1. I don't think there's a need for `torch.no_grad()` here - we should be able to run in the default torch context?
2. I'm not sure this necessarily works for initialising the modules - suppose that `WhisperEncoder` comes before the embedding layer in the list when we loop over all the modules: we first set the embeddings to sinusoids, but then set it to random again when we get to the embeddings

cc @ArthurZucker - do you know of a clean way to initialise a particular `nn.Embedding` layer with sinusoids, but not touch others?

Reply:
1. It's a standard practice to use `torch.no_grad()` for weight init functions, since we don't want to capture computation graph of weight init. See https://github.com/pytorch/pytorch/blob/dac895c10ae0af86a5ca1183c1d08d0c00bc6c43/torch/nn/init.py#L17-L19 for example. However, in this case, since the sinusoidal embeddings have `.requires_grad = False`, it is possible to not use `torch.no_grad()` here. But I still prefer to have no grad for weight init.
2. As I stated above, `nn.Module.apply()` is depth-first search: the embedding layer (leaf module) will always come before `WhisperEncoder` (unless PyTorch changes this behaviour in the future, which is highly unlikely). 

I suppose to make it bullet-proof, the most straight-forward way is to set a private attribute. e.g.

```python
# in WhisperEncoder.__init__()
        self.embed_positions = nn.Embedding(self.max_source_positions, embed_dim)
        self.embed_positions.requires_grad_(False)
        self.embed_positions._hf_sinusoids_init = True

# in _init_weights()
        elif isinstance(module, nn.Embedding):
            if getattr(module, "_hf_sinusoids_init", False):
                # sinusoids init
            else:
                # normal embedding init
```

Reply:
Okay cool - thanks for the detailed explanations @gau-nernst! Good with me to keep as is then!